### PR TITLE
Marketplace API - swagger for version 2018-01-31

### PR DESCRIPTION
**Pull request by kjeur**
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Comment by lmazuel** [comment]
@kjeur While investigating the generated code from that PR, I saw an incorrect declaration of `budgetNameParameter` that should be declared as scope to method, and not global. Please take advantage of this PR to add this to `budgetNameParameter`
`"x-ms-parameter-location": "method"`

**Comment by kjeur** [comment]
@lmazuel Thanks! Do you want me to copy this to all operations? This parameter is being used in GET, PUT and DELETE operation.

**Comment by lmazuel** [comment]
@kjeur It's a one time fix at the end of the file actually :)
```json
    "budgetNameParameter": {
      "name": "name",
      "in": "path",
      "description": "Budget name.",
      "required": true,
      "type": "string",
      "x-ms-parameter-location": "method"
    }
```

**Comment by kjeur** [comment]
@lmazuel That's leaning for me. Thanks!

**Comment by lmazuel** [comment]
@kjeur Seems to me that new ApiVersion is isofunctional with 2017-11-30, with just new operation group. Could you confirm? If so, I would suggest python package 1.2.0, and I will update your Python conf to reflect this version number.

**Comment by sandeepnl** [comment]
@lmazuel Yes, we can have python package version as 1.2.0. But let us not start on the python package yet after this PR since other devs are actively working on adding more operations to the same <swagger,api-version>. I think once its all done, we can publish a single package with version 1.2.0

**Comment by lmazuel** [comment]
Thanks @sandeepnl ! I'm just adding the info in your Readme, so automatic Python PR are ready when you think it's time.

**Comment by sandeepnl** [comment]
@lmazuel The error [here](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/333414715) is an ongoing issue as per this [bug](https://github.com/Azure/oav/issues/197). Please confirm

**Comment by lmazuel** [comment]
@sandeepnl @kjeur Yes, seems to me it's this one. You should be able to fix all the other errors however.
@amarzavery @jhendrixMSFT @mcardosos What's the recommendation when CI fails for a OVA bug?

**Comment by mcardosos** [comment]
- @amarzavery or I can take a look at the bug.
- If it is a bug, share enough info about the bug and open an issue on the OAV repo.
- If we don't get a solution fast enough, you can decide if the change gets in or not yet.

Would you add something, Amar?

**Comment by sandeepnl** [comment]
@lmazuel looks like all the errors in that leg are due to the same bug, because we dont see any problem n the definition. Or are we missing something?

File: specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/consumption.json [hunk 1]
```diff
@@ -117,6 +117,79 @@
         }
       }
     },
+    "/{scope}/providers/Microsoft.Consumption/marketplaces": {
+      "get": {
+        "tags": [
+          "Marketplaces"
+        ],
+        "operationId": "Marketplaces_List",
```
Comment (mcardosos) [inline specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/consumption.json]: [Odata extension](https://github.com/Azure/autorest/blob/master/docs/extensions/readme.md#x-ms-odata) could be useful here :D

File: specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/consumption.json [hunk 2]
```diff
@@ -672,6 +745,170 @@
...
+      ],
+      "properties": {
+        "properties": {
+          "x-ms-client-flatten": true,
+          "$ref": "#/definitions/MarketplaceProperties",
+          "title": "Marketplace properties"
+        }
+      }
+    },
+    "MarketplacesListResult": {
```
Comment (mcardosos) [inline specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/consumption.json]: It is a good idea to explicitly express this is an object with `"type": "object"`

File: specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/consumption.json [hunk 3]
```diff
@@ -672,6 +745,170 @@
...
+          }
+        },
+        "nextLink": {
+          "description": "The link (url) to the next page of results.",
+          "type": "string",
+          "readOnly": true
+        }
+      }
+    },
+    "MarketplaceProperties": {
```
Comment (mcardosos) [inline specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/consumption.json]: Same here, adding `"type": "object"` is a good idea.

**Comment by lmazuel** [comment]
@sandeepnl OAV109 is the bug you pointed, but there is two OAV108 issues that does not seem related IMHO

**Review by mcardosos — CHANGES_REQUESTED** [review]
Some of this observations might get rid of oav errors.

**Comment by sandeepnl** [comment]
@lmazuel yes all those oav errors seems cryptic frankly we are not understanding what it is complaining about. Can someone help us filter the information on what the error is?

File: specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/consumption.json (hunk 1 above)
Comment (kjeur) [reply specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/consumption.json]: @mcardosos done.

File: specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/consumption.json (hunk 2 above)
Comment (kjeur) [reply specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/consumption.json]: @mcardosos done

**Comment by kjeur** [comment]
@mcardosos suggested changes are pushed.

**Comment by mcardosos** [comment]
@kjeur err, I dont see them

**Comment by kjeur** [comment]
@mcardosos My bad, push was failed. Check now.

File: specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/examples/MarketplacesList.json [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+              "billingPeriodId": "/subscriptions/subid/providers/Microsoft.Billing/billingPeriods/201702",
+              "usageStart": "2017-02-13T00:00:00Z",
+              "usageEnd": "2017-02-13T23:59:59Z",
+              "instanceName": "shared1",
+              "instanceId": "/subscriptions/subid/resourceGroups/Default-Web-eastasia/providers/Microsoft.Web/sites/shared1",
+              "currency": "USD",
+              "consumedQuantity": 0.00328,
+              "pretaxCost": 0.67,
+              "isEstimated": false,
+              "offerName": "offer1",
```
Comment (mcardosos) [inline specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/examples/MarketplacesList.json]: So this case it is not an OAV bug. According to the swagger, it looks like this should be `offerType` instead

File: specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/examples/MarketplacesListForBillingPeriod.json [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+              "usageStart": "2017-02-13T00:00:00Z",
+              "usageEnd": "2017-02-13T23:59:59Z",
+              "instanceName": "shared1",
+              "instanceId": "/subscriptions/subid/resourceGroups/Default-Web-eastasia/providers/Microsoft.Web/sites/shared1",
+              "currency": "USD",
+              "consumedQuantity": 0.00328,
+              "pretaxCost": 0.67,
+              "isEstimated": false,
+              "meterId": "fe167397-a38d-43c3-9bb3-8e2907e56a41",
+              "offerName": "offer1",
```
Comment (mcardosos) [inline specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/examples/MarketplacesListForBillingPeriod.json]: It is `offerType`

**Review by mcardosos — CHANGES_REQUESTED** [review]
Found model validator issues

File: specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/examples/MarketplacesListForBillingPeriod.json (hunk 5 above)
Comment (kjeur) [reply specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-01-31/examples/MarketplacesListForBillingPeriod.json]: @mcardosos renamed it. Should be offerName, changed in mode.

**Comment by mcardosos** [comment]
@kjeur @sandeepnl is this already available?

**Comment by kjeur** [comment]
@mcardosos Yes.

**Review by mcardosos — APPROVED** [review]
Cool!
Thanks for the awesome teamwork :D 
Just for awareness, there are still [model validation errors](https://travis-ci.org/Azure/azure-rest-api-specs/jobs/333522265#L728), but since they were not introduced in this PR, can be fixed later.
Also, since this is a new operation, @ravbhatnagar should also take a look

**Comment by ms-premp** [comment]
@ravbhatnagar please have a look

**Comment by ravbhatnagar** [comment]
Reviewed this in person. Signing off!

**Pull request merged**